server/datastore/mysql/software_titles.go

```diff
@@ -125,77 +127,198 @@ SELECT
-		whereClause = "h.team_id = ?"
-		args = append(args, opt.TeamID)
+		args[0] = *opt.TeamID
 	}
 
-	if match := opt.ListOptions.MatchQuery; match != "" {
-		whereClause += " AND (st.name LIKE ? OR scve.cve LIKE ?)"
+	additionalWhere := ""
+	match := opt.ListOptions.MatchQuery
+	softwareJoin := ""
+	if match != "" || opt.VulnerableOnly {
+		softwareJoin = fmt.Sprintf(`
+			JOIN software s ON s.title_id = st.id
+			-- placeholder for changing the JOIN type to filter vulnerable software
+			%s JOIN software_cve scve ON s.id = scve.software_id
+		`, cveJoinType)
+	}
+
+	if match != "" {
+		additionalWhere += " AND (st.name LIKE ? OR scve.cve LIKE ?)"
 		match = likePattern(match)
 		args = append(args, match, match)
 	}
 
-	stmt = fmt.Sprintf(stmt, cveJoinType, hostsJoin, whereClause)
+	stmt = fmt.Sprintf(stmt, softwareJoin, additionalWhere)
 	return stmt, args
 }
 
 func selectSoftwareVersionsSQL(titleIDs []uint, teamID uint, withCounts bool) (string, []any, error) {
 	selectVersionsStmt := `
 SELECT
-	st.id as title_id,
+	s.title_id,
 	s.id, s.version,
 	%s -- placeholder for optional host_counts
 	CONCAT('[', GROUP_CONCAT(JSON_QUOTE(scve.cve) SEPARATOR ','), ']') as vulnerabilities
-FROM software_titles st
-JOIN software s ON s.title_id = st.id
-LEFT JOIN host_software hs ON hs.software_id = s.id
-LEFT JOIN software_cve scve ON s.id = scve.software_id
-%s -- placeholder for optional JOIN ON host_counts
-WHERE st.id IN (?)
+FROM software s
+LEFT JOIN software_host_counts shc ON shc.software_id = s.id
+LEFT JOIN software_cve scve ON shc.software_id = scve.software_id
+WHERE s.title_id IN (?)
+AND shc.team_id = ?
+AND shc.hosts_count > 0
 GROUP BY s.id`
 
-	var args []any
 	extraSelect := ""
-	extraJoin := ""
 	if withCounts {
-		args = append(args, teamID)
 		extraSelect = "MAX(shc.hosts_count) AS hosts_count,"
-		extraJoin = `
-			JOIN software_host_counts shc
-			ON shc.software_id = s.id
-				AND shc.hosts_count > 0
-				AND shc.team_id = ?
-		`
 	}
 
-	args = append(args, titleIDs)
-	selectVersionsStmt = fmt.Sprintf(selectVersionsStmt, extraSelect, extraJoin)
-	selectVersionsStmt, args, err := sqlx.In(selectVersionsStmt, args...)
+	selectVersionsStmt = fmt.Sprintf(selectVersionsStmt, extraSelect)
+	selectVersionsStmt, args, err := sqlx.In(selectVersionsStmt, titleIDs, teamID)
 	if err != nil {
 		return "", nil, fmt.Errorf("bulding sqlx.In query: %w", err)
 	}
 	return selectVersionsStmt, args, nil
 }
+
+// SyncHostsSoftwareTitles calculates the number of hosts having each
+// software installed and stores that information in the software_titles_host_counts
+// table.
+func (ds *Datastore) SyncHostsSoftwareTitles(ctx context.Context, updatedAt time.Time) error {
+	const (
+		resetStmt = `
+            UPDATE software_titles_host_counts
+            SET hosts_count = 0, updated_at = ?`
+
+		globalCountsStmt = `
+            SELECT
+                COUNT(DISTINCT hs.host_id),
+                0 as team_id,
```

Comment:
One thing to remember, if we ever want counts for "no team" distinct from "all teams", we'll need some other way to store those stats (nothing to do now, just a thought I had seeing how we store those counts).

Reply:
agreed, I was on the fence about doing the change now, or tackling both software and software titles when we inevitably decide to do that.

what do you think? I can tweak this if this will save us a headache

Reply:
I wouldn't do it until we have a need for "no team", keeping this PR focused on just its goal.